api/agent/agent.go

```diff
@@ -556,7 +460,29 @@ func (s *hotSlot) exec(ctx context.Context, call *call) error {
 	errApp := make(chan error, 1)
 	go func() {
 		ci := protocol.NewCallInfo(call.Call, call.req)
-		errApp <- proto.Dispatch(ctx, ci, call.w)
+		err := proto.Dispatch(ctx, ci, call.w)
+
+		// TODO: narrow this down to StatusBadGateway
+		// TODO: we can skip this for Cold containers, but let's not optimize Cold path
+		if err != nil && models.IsAPIError(err) {
+			// attempt to drain stdout before it gets closed.
+			// Typical example case is when we encounter a parse error upon read from container. eg. corrupt
+			// json or bad http. This typically means, the stdout has more data that has not been read, which
+			// will trip next request.
+			done := make(chan struct{})
+			go func() {
+				io.CopyN(ioutil.Discard, stdoutRead, 100*1024*1024)
+				close(done)
+			}()
+
+			// let drain, 50 msecs or 100MB, whichever comes first
+			select {
+			case <-done:
+			case <-time.After(time.Duration(50) * time.Millisecond):
+			}
```

Comment:
we seem to have hit 3 cases where it makes a lot more sense to just shut the container down than try to recover. i give up, let's do that. killing off the ghost thing might be nice, but it means we end up with a weird stderr pipe too, anyway, an option

Reply:
yup, I agree... (Also I know this is not going to interact nice with I'm going to merge in the next few days: I/O size clamping)